[pytorch][PR] Add ability for a mobile::Module to save as flatbuffer

qihqi opened the pull request:
Summary:
Included functions:
save_mobile_module -> saves a mobile::Module to flatbuffer
load_mobile_module_from_file -> loads a flatbuffer into mobile::Module
parse_mobile_module -> parses from bytes or deserialized flatbuffer module object

Compared to previous attempts, this diff only adds flatbuffer to cmake target and leaves fbcode/xplat ones unchanged.

Test Plan: unittest

Differential Revision: D33239362

qihqi commented:
@pytorchbot ciflow rerun -l ciflow/all

gmagogsfm commented on torch/csrc/jit/serialization/mobile_bytecode.fbs:
```diff
@@ -0,0 +1,197 @@
+}
+
+table Module {
+  version:int;
```
What is this version supposed to be?

qihqi replied:
1 + whatever the latest (and last) pickle version should be. It is not explicitly used now.

gmagogsfm commented on torch/csrc/jit/serialization/flatbuffer_serializer.cpp:
```diff
@@ -0,0 +1,681 @@
+                 ival_pos)
+                 .Union();
+  } else {
+    AT_ERROR("Unknown IValue type for pickling: ", ivalue.tagKind());
```
LOL, I think we are done with "pickling"

qihqi replied:
rephased

qihqi commented on torch/csrc/jit/serialization/flatbuffer_serializer.cpp:
```diff
@@ -0,0 +1,681 @@
+      flatbuffers::FlatBufferBuilder& fbb,
+      ClassTypePtr class_type);
+
+  // cached stuff
```
removed.

qihqi commented on torch/csrc/jit/serialization/flatbuffer_serializer.cpp:
```diff
@@ -0,0 +1,681 @@
+    }
+  } catch (const std::runtime_error&) {
+  } catch (const c10::Error&) {
+  }
```
Added more comments. IValue::hash explicitly throws runtime_error for some IValues. And for Tensor operator== doesn't return bool so you get c10::Error.

qihqi commented on torch/csrc/jit/serialization/flatbuffer_serializer.cpp:
```diff
@@ -0,0 +1,681 @@
+    flatbuffers::FlatBufferBuilder& fbb,
+    const IValue& ivalue) {
+  if (ivalue.isNone()) {
+    return 0;
```
done

malfet commented on .jenkins/pytorch/win-build.sh:
```diff
@@ -56,6 +56,8 @@ if [[ $PYLONG_API_CHECK == 0 ]]; then
 fi
 set -ex
 
+echo 'python %*' > /c/Windows/py.bat
+
```
Can you please explain, why this change is needed?

qihqi replied:
So flatbuffers CMakeList.txt calls the 'py' command which is not available. Discussed internally: https://fb.workplace.com/groups/1274980366011614/permalink/2066627280180248/

malfet replied:
Can you please share a reference to the source code in flatbuffers that calls that?

qihqi replied:
https://github.com/google/flatbuffers/blob/master/CMakeLists.txt#L533

malfet commented on BUILD.bazel:
```diff
@@ -1884,15 +1884,17 @@ cc_test(
+  "test/cpp/tensorexpr/*.cpp",
+  "test/cpp/tensorexpr/*.h",
+])
+jit_test_source.remove("test/cpp/jit/test_flatbuffer.cpp")
```
Can you please add a comment, why this test should not be run in OSS CI?

qihqi replied:
Because this PR's goal is to add flatbuffer to CMake's target only. Previous attempts add flatbuffer to cmake/bazel/fbcode's TARGET/xplat's BUCK at the same time and had bigger impact surfaces. So we opted to add those in separate PR's so each can be reverted individually if something unexpected happens.

malfet replied:
This sounds wrong. You can make this an internal only feature initially, but I would prefer no feature discrepancy between CMAKE and Bazel

malfet commented on CMakeLists.txt:
```diff
@@ -775,6 +775,7 @@ if(NOT MSVC)
   endif()
   if(CMAKE_COMPILER_IS_GNUCXX AND NOT (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 7.0.0))
     string(APPEND CMAKE_CXX_FLAGS " -Wno-stringop-overflow")
+    string(APPEND CMAKE_CXX_FLAGS " -Wno-noexcept-type")
```
Again, any explanation why this is needed?

qihqi replied:
So gcc has a warning telling us that functions with noexcept annotation will have mangled name changed in C++17: https://stackoverflow.com/questions/46798456/handling-gccs-noexcept-type-warning

And since we have -Werror that becames an Error. I got it as a CI failure in previous PRs on files that I didnt modify which I don't know why my PR tickled.

I'll try removing it now since HEAD moved quite a bit to see if any failure shows up.

qihqi replied:
removed.

malfet commented on scripts/gen_flatbuffer.sh:
```diff
@@ -0,0 +1,24 @@
+mkdir build
+cd build || exit
+py() { command python "$@"; }
+if [[ "$OSTYPE" == "darwin"* ]]; then
+   cmake -G "Xcode" -DCMAKE_BUILD_TYPE=Release -DCMAKE_OSX_ARCHITECTURES=x86_64 ..
+   cmake --build . --target flatc
+   # for some reason on mac the binary appears in a different directory...
+   FLATC=./Debug/flatc
+else
```
Can you please explain why Darwin build needs to be special-cased? This would not work on Darwin systems where only Xcode tools are installed rather than xcode. And what is wrong with regular `make`/`ninja` build? It would be nice to pass generator user selected when running top level CMAKE? (Should be deinfed as top level `CMAKE_GENERATOR` variable)

qihqi replied:
I copied the build command from here: https://google.github.io/flatbuffers/flatbuffers_guide_building.html

Q: What is the different between`-G "Xcode"` and `-G "xcode"`?

Q2: Along with previous comment should I change it to `-G $CMAKE_GENERATOR`?

As for why it appears as ./Debug/flatc, I don't know... I got CI failure on file not found then I printed out generated files and saw the path in ./Debug/flatc. If you know how to control this please let me know.

malfet reviewed (changes requested):
Please modify `gen_flatbuffer.sh` to respect callee choice of CMAKE_GENERATOR and also work correctly with cross compilation on Darwin (i.e. please build flatc as dual-archi x86 and arm64 binary)

Also, please specify in PR comment what version of flatbuffer is imported as dependency (it's preferably to synchronize checkout commit to one of the release tags)

malfet commented on torch/csrc/jit/serialization/mobile_bytecode.fbs:
```diff
@@ -0,0 +1,197 @@
+
+table TensorMetadata {
+  // torch._utils _rebuild_tensor_v2
+  storage_location_index: uint;
```
I don't know much about fbs schema, but looks like this file has an inconsistent with regard to whitespace after colon. In this table it is used, but say in QuantizedSchema there are no spaces between name and type of the field
```suggestion
  storage_location_index:uint;
```

qihqi replied:
done.

qihqi commented:
> Please modify `gen_flatbuffer.sh` to respect callee choice of CMAKE_GENERATOR and also work correctly with cross compilation on Darwin (i.e. please build flatc as dual-archi x86 and arm64 binary)

@malfet Can I interpret this suggestion as use `cmake -G $CMAKE_GENERATOR` when building `flatc`?

I don't think we can flatc for arm64 though because it fails like this: https://github.com/pytorch/pytorch/runs/4411131194?check_suite_focus=true#step:6:4427

Also why should we build flatc for arm64 if we only ever run it on the host machine (x86) and not arm64?

qihqi commented:
Things I tried:
* use `-G "$CMAKE_GENERATOR"` when building flatc. linux build fails with `CMake Error: Could not create named generator ` (https://pipelines.actions.githubusercontent.com/mBh68xKhi8LyM7tp3vECvYXNFvuV4gyVGgmYCteuEZP9JH92QN/_apis/pipelines/1/runs/1360792/signedlogcontent/8?urlExpires=2021-12-22T17%3A06%3A26.9827936Z&urlSigningMethod=HMACV1&urlSignature=pU9oiYAGm5lHh3bBo90kNsoKUQHGpxkVoUME4L7gHWI%3D)

* leave -G out for both linux and macos, then fails with `Unsupported architecture` because flatc doesn't build for arm64 (https://pipelines.actions.githubusercontent.com/mBh68xKhi8LyM7tp3vECvYXNFvuV4gyVGgmYCteuEZP9JH92QN/_apis/pipelines/1/runs/1362199/signedlogcontent/8?urlExpires=2021-12-22T17%3A57%3A15.3624937Z&urlSigningMethod=HMACV1&urlSignature=nTwtlujDEglm7cz5MmlNvS4ud4LJZMchku0Do5dQQpE%3D). Presumably leaving `-G "Xcode"` out makes `-DCMAKE_OSX_ARCHITECTURES=x86_64` not working

qihqi commented:
Hi @malfet any other ideas that I should try before landing? I left the few failed CIs there in case you want to look at the logs. If not I'll reverted back to my last known state that passed all CIs. Thanks!

malfet commented on scripts/gen_flatbuffer.sh:
```diff
@@ -0,0 +1,22 @@
+mkdir build
+cd build || exit
+py() { command python "$@"; }
+cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_OSX_ARCHITECTURES=x86_64 ..
```
Please compile flatbuffers as fat binary for MacOS, also as suggested in initial review, re-use CMAKE_GENERATOR from the parent project
```suggestion
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_OSX_ARCHITECTURES=x86_64,arm64  ..
```

malfet commented on scripts/gen_flatbuffer.sh:
```diff
@@ -0,0 +1,22 @@
+py() { command python "$@"; }
+cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_OSX_ARCHITECTURES=x86_64 ..
+cmake --build . --target flatc
+if [[ "$OSTYPE" == "darwin"* ]]; then
+   # for some reason on mac the binary appears in a different directory...
+   FLATC=./Debug/flatc
+else
```
If `CMAKE_GENERATOR` is reused, it should not be necessary (or this condition should be `if [[ $CMAKE_GENERATOR == Xcode ]];`

qihqi commented:
Sync'd offline with @malfet and agreed to check-in the generated header file instead of messing with build scripts to generate it on the fly. PTAL.

gmagogsfm commented:
> Sync'd offline with @malfet and agreed to check-in the generated header file instead of messing with build scripts to generate it on the fly. PTAL.

LGTM

malfet commented on .jenkins/pytorch/win-build.sh:
```diff
@@ -56,6 +56,8 @@ if [[ $PYLONG_API_CHECK == 0 ]]; then
 fi
 set -ex
 
+echo 'python %*' > /c/Windows/py.bat
```
As headers are checked in, this should not longer be necessary, should it?

malfet commented on torch/CMakeLists.txt:
```diff
@@ -70,6 +70,7 @@ set(TORCH_PYTHON_INCLUDE_DIRECTORIES
 
     ${TORCH_ROOT}/third_party/gloo
     ${TORCH_ROOT}/third_party/onnx
+    ${TORCH_ROOT}/third_party/flatbuffers/include
```
LGTM, but in followup PR can we limit it to just 2 files in question? (So that we don't leak flatbuffers dependency to module extensions)

qihqi replied:
Do you mean we can use different include paths when compiling `flatbuffer_loader.cpp` and `flatbuffer_serializer.cpp` but not when compiling other .cpp files? How do I do that? Also it's not just those 2 files also ones that happens to include `flatbuffer_loader.h` or `flatbuffer_serializer.h`.

malfet commented on torch/CMakeLists.txt:
```diff
@@ -345,6 +346,8 @@ if(HAVE_SOVERSION)
       VERSION ${TORCH_VERSION} SOVERSION ${TORCH_SOVERSION})
 endif()
 add_dependencies(torch_python torch_python_stubs)
+add_dependencies(torch_python flatbuffers)
```
Hmm, I though flatbuffers is header only ones its generated, isn't it?

malfet commented:
Looks good to me, but please remove changes to windows-builds (they should no longer be necessary)
Also, can you please add to PR description projected binary size increase as result of this change?
And why can't it be an optional feature?

qihqi committed:
[pytorch][PR] Add ability for a mobile::Module to save as flatbuffer (#70201)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/70201

Included functions:
save_mobile_module -> saves a mobile::Module to flatbuffer
load_mobile_module_from_file -> loads a flatbuffer into mobile::Module
parse_mobile_module -> parses from bytes or deserialized flatbuffer module object

Compared to previous attempts, this diff only adds flatbuffer to cmake target and leaves fbcode/xplat ones unchanged.

Test Plan: unittest

Reviewed By: malfet, gmagogsfm

Differential Revision: D33239362

fbshipit-source-id: c89d5f8b96fee6906a2df9f034c081865b34548f

qihqi commented:
> Looks good to me, but please remove changes to windows-builds (they should no longer be necessary) Also, can you please add to PR description projected binary size increase as result of this change? And why can't it be an optional feature?

Update: removed change to win-build.sh and it seems to work. Binary size increase is ~15kb. We didn't propose as optional feature because this will replace Module._save_for_lite_interpreter().

qihqi commented:
> Looks good to me, but please remove changes to windows-builds (they should no longer be necessary) Also, can you please add to PR description projected binary size increase as result of this change? And why can't it be an optional feature?

Update: removed change to win-build.sh and it seems to work. Binary size increase is ~15kb. We didn't propose as optional feature because this will replace Module._save_for_lite_interpreter().
